vault/token_store.go

```diff
@@ -1123,27 +1124,31 @@ func (ts *TokenStore) revokeSalted(ctx context.Context, saltedID string) (ret er
 		return err
 	}
 
-	// If we are returning an error, mark the entry with -3 to indicate
-	// failed revocation. This way we don't try to clean up during active
-	// revocation (-2).
 	defer func() {
-		if ret != nil {
+		if deleteEntry || ret != nil {
 			lock.Lock()
 			defer lock.Unlock()
+		}
 
-			// Lookup the token again to make sure something else didn't
-			// revoke in the interim
-			entry, err := ts.lookupSalted(ctx, saltedID, true)
```

Comment:
I've been thinking about this, and I think removing this is okay. Marking useCount as -2 is in a protected critical path, so the only way we should be here is if the token was set to -2 by this goroutine.

Reply:
Although this actually begs another question: do we even need to grab the lock here, then? Lookup won't return the token if use count is negative, so locking when the Delete happens shouldn't matter. Nothing else should be operating on it because at that point useCount is -2 so any other revocation will skip being run. If the Delete fails, we update the entry with -3, but the timing of that also doesn't matter -- anything that tries to revoke before that won't work (it'll hit the -2), and anything that tries after will keep going, but this goroutine will have exited.

Reply:
I think we need to still grab the lock since the `entry.NumUses == tokenRevocationInProgress` check happens before the write. There could theoretically be a case where 2 goroutines calling revokeSalted have checked that `entry.NumUses` is not -2 before either one got to update it, and so the revokeSalted operation would go through twice.

```
	// On failure we write -3, so if we hit -2 here we're already running a
	// revocation operation. This can happen due to e.g. recursion into this
	// function via the expiration manager's RevokeByToken.
```

I wonder if we need `tokenRevocationInProgress` since `RevokeByToken` does not remove the actual token due to more recent changes (`m.revokeCommon(tokenLeaseID, false, true)` in [here](https://github.com/hashicorp/vault/blob/78bb9d777b61723958595c6e74b940c7039e2c3f/vault/expiration.go#L590)) so the recursion in the comment above might no longer be an issue.